File: companion/quest/Assets/Scripts/SocketClient.cs
// Copyright (c) Meta Platforms, Inc. and affiliates.
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the root directory of this source tree.

using System;
using System.Collections.Generic;
using SocketIOClient;
using SocketIOClient.JsonSerializer;
using UnityEngine;

namespace HapticStudio
{
    /// <summary>
    /// Socket client that wraps the SocketIO dependency
    /// </summary>
    public class SocketClient
    {
        public SocketEndpoint Endpoint => _endpoint;

        private SocketIOUnity _socket;
        protected SocketEndpoint _endpoint;
        protected bool _isPingSocket;

        /// <summary>
        /// Initializer.
        /// </summary>
        /// <param name="endpoint">The socket endpoint</param>
        /// <param name="ping">Flag that marks the socket as a ping socket, used to check for server availability.</param>
        public SocketClient(SocketEndpoint endpoint, bool ping = false)
        {
            _endpoint = endpoint;
            _isPingSocket = ping;
            var uri = new Uri($"ws://{endpoint.ip}:{endpoint.port}");

            _socket = new SocketIOUnity(uri, new SocketIOOptions
            {
                Query = new Dictionary<string, string>
                    {
                        { "deviceId", SystemInfo.deviceUniqueIdentifier },
                        { "name", SystemInfo.deviceName },
                        { "model", SystemInfo.deviceModel },
                        { "version", Application.version },
                        { "ping", ping.ToString() }
                    },
                // @oss-disable
                // EIO = EngineIO.V4, // @oss-enable
                Path = "/ws",
                Reconnection = !ping,
                Transport = SocketIOClient.Transport.TransportProtocol.WebSocket
            });
        }

        public virtual void Connect()
        {
            if (_endpoint == null) return;

            if (!_isPingSocket)
            {
                Debug.Log($"Connecting to ws://{_endpoint.ip}:{_endpoint.port}");
            }

            _socket.Connect();
        }

        public virtual void Disconnect()
        {
            _socket.Disconnect();
        }

        public IJsonSerializer JsonSerializer { get => _socket.JsonSerializer; set => _socket.JsonSerializer = value; }

        public virtual bool Connected { get => _socket.Connected; }

        public virtual event EventHandler OnConnected { add { _socket.OnConnected += value; } remove { _socket.OnConnected -= value; } }
        public virtual event EventHandler<string> OnError { add { _socket.OnError += value; } remove { _socket.OnError -= value; } }
        public virtual event EventHandler<string> OnDisconnected { add { _socket.OnDisconnected += value; } remove { _socket.OnDisconnected -= value; } }

        public virtual void On(string eventName, Action<SocketIOResponse> callback)
        {
            _socket.On(eventName, callback);
        }

        public virtual void Emit(string eventName, params object[] data)
        {
            _socket.Emit(eventName, data);
        }

        public virtual void EmitStringAsJSON(string eventName, string json)
        {
            _socket.EmitStringAsJSON(eventName, json);
        }
    }
}

File: companion/android/app/src/main/java/com/example/hapticsstudio/net/SocketClient.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.example.hapticsstudio.net

import android.util.Log
import com.example.hapticsstudio.model.SocketEndpoint
import io.socket.client.IO
import io.socket.client.Socket
import io.socket.emitter.Emitter
import org.json.JSONObject

private const val TAG = "SocketClient"

/**
 * Thin wrapper over the official `io.socket` [Socket] client. Connects to the desktop's Socket.IO
 * server (`http://<ip>:<port>`, path `/ws`) over the websocket transport, forwarding the device
 * identity as connection query parameters.
 */
class SocketClient {

  private var socket: Socket? = null

  fun connect(endpoint: SocketEndpoint, query: String, onReady: () -> Unit = {}) {
    disconnect()

    val url = "http://${endpoint.ip}:${endpoint.port}"
    val options =
        IO.Options().apply {
          path = "/ws"
          transports = arrayOf("websocket")
          reconnection = true
          reconnectionDelay = 1000
          this.query = query
        }

    val client =
        try {
          IO.socket(url, options)
        } catch (e: Exception) {
          Log.e(TAG, "Failed to create socket for $url", e)
          return
        }

    socket = client
    // Register application handlers on the live socket BEFORE connecting so no early events
    // (e.g. auth_required / auth_granted) are missed.
    onReady()
    client.connect()
  }

  fun on(event: String, listener: Emitter.Listener) {
    socket?.on(event, listener)
  }

  fun emit(event: String, payload: JSONObject) {
    socket?.emit(event, payload)
  }

  fun emit(event: String) {
    socket?.emit(event)
  }

  fun disconnect() {
    socket?.let {
      it.off()
      it.disconnect()
      it.close()
    }
    socket = null
  }
}
